pytorch_lightning/strategies/bagua.py

```diff
@@ -0,0 +1,235 @@
+class LightningBaguaModule(_LightningModuleWrapperBase):
+    def __init__(self, pl_module: "pl.LightningModule") -> None:
+        super().__init__(pl_module)
+        # Bagua use `bagua_module_name` to distinguish different modules
+        self._bagua_module_name = pl_module._get_name() + str(id(pl_module))
+
+
+_bagua_reduce_ops: Dict[ReduceOp, BaguaReduceOp] = {}
+
+
+def from_torch_reduce_op(op: ReduceOp) -> Optional[BaguaReduceOp]:
+    """Convert a `torch.distributed.ReduceOp` to its equivalent `bagua.torch_api.ReduceOp`."""
+    global _bagua_reduce_ops
+
+    if len(_bagua_reduce_ops) == 0:
+        _bagua_reduce_ops.update(
+            {
+                ReduceOp.SUM: BaguaReduceOp.SUM,
+                ReduceOp.PRODUCT: BaguaReduceOp.PRODUCT,
+                ReduceOp.MIN: BaguaReduceOp.MIN,
+                ReduceOp.MAX: BaguaReduceOp.MAX,
+                ReduceOp.BAND: BaguaReduceOp.BAND,
+                ReduceOp.BOR: BaguaReduceOp.BOR,
+                ReduceOp.BXOR: BaguaReduceOp.BXOR,
+            }
+        )
+
+    return _bagua_reduce_ops.get(op, None)
+
+
+class BaguaStrategy(DDPStrategy):
+    distributed_backend = _StrategyType.BAGUA
+
+    def __init__(
+        self,
+        algorithm: str = "gradient_allreduce",
+        do_flatten: bool = True,
+        accelerator: Optional["pl.accelerators.accelerator.Accelerator"] = None,
+        parallel_devices: Optional[List[torch.device]] = None,
+        cluster_environment: Optional[ClusterEnvironment] = None,
+        checkpoint_io: Optional[CheckpointIO] = None,
+        precision_plugin: Optional[PrecisionPlugin] = None,
+        **kwargs: Union[Any, Dict[str, Any]],
+    ):
+        """Provides capabilities to run distributed training using the `Bagua <https://github.com/BaguaSys/bagua>`_
+        library.
+
+        Arguments:
+            algorithm:  Distributed algorithm used to do the actual communication and update. Built-in algorithms
+                include "gradient_allreduce", "bytegrad", "decentralized", "low_precision_decentralized", "qadam" and
+                "async".
+            do_flatten: Whether to flatten the Bagua communication buckets. The flatten operation will reset data
+                pointer of bucket tensors so that they can use faster code paths. Default: ``True``.
+            kwargs: Additional arguments that will be passed to initialize the Bagua algorithm.
+        """
+
+        super().__init__(
+            accelerator=accelerator,
+            parallel_devices=parallel_devices,
+            cluster_environment=cluster_environment,
+            checkpoint_io=checkpoint_io,
+            precision_plugin=precision_plugin,
+        )
+
+        self._bagua_algorithm = algorithm
+        self._bagua_do_flatten = do_flatten
+        self._bagua_kwargs = kwargs
+
+    def setup_environment(self) -> None:
+        # start the other scripts
+        if not self.cluster_environment.creates_processes_externally:  # type: ignore[union-attr]
```

Comment:
n00b question, this is only called when we do not use the bagua launcher right?

Reply:
Bagua launcher will not call this. It seems other env like `TorchElasticEnvironment`, `KubeflowEnvironment`  and `LSFEnvironment` will not call it either, and I have not tested for those. Have tested for `LightningEnvironment`, this will be called with a condition.